vpr/src/pack/prepack.h

```diff
@@ -6,16 +6,175 @@
+ * // ...
+ * prepacker.reset(); // Or if the prepacker object is destroyed.
+ * // Prepacked molecules can no longer be used beyond this point.
+ *
+ */
+class Prepacker {
+    using atom_molecules_const_iterator = std::multimap<AtomBlockId, t_pack_molecule*>::const_iterator;
+public:
+    // The constructor is default, the init method performs prepacking.
+    Prepacker() = default;
+
+    // This class maintains pointers to internal data structures, and as such
+    // should not be copied or moved (prevents unsafe accesses).
+    Prepacker(const Prepacker&) = delete;
+    Prepacker& operator=(const Prepacker&) = delete;
+
+    /**
+     * @brief Performs prepacking.
+     *
+     * Initializes the prepacker by performing prepacking and allocating the
+     * necessary data strucutres.
+     *
+     *  @param atom_nlist           The atom netlist to prepack.
+     *  @param logical_block_types  A list of the logical block types on the device.
+     */
+    void init(const AtomNetlist& atom_nlist, const std::vector<t_logical_block_type> &logical_block_types);
+
+    /**
+     * @brief Get the cluster molecules containing the given atom block.
+     *
+     *  @param blk_id       The atom block to get the molecules of.
+     */
+    inline vtr::Range<atom_molecules_const_iterator> get_atom_molecules(AtomBlockId blk_id) const {
+        auto rng = atom_molecules.equal_range(blk_id);
+        return vtr::make_range(rng.first, rng.second);
+    }
+
+    /**
+     * @brief Get the expected lowest cost physical block graph node for the
+     *        given atom block.
+     *
+     *  @param blk_id       The atom block to get the pb graph node of.
+     */
+    inline t_pb_graph_node* get_expected_lowest_cost_pb_gnode(AtomBlockId blk_id) const {
+        auto iter = expected_lowest_cost_pb_gnode.find(blk_id);
+        VTR_ASSERT(iter != expected_lowest_cost_pb_gnode.end());
+        t_pb_graph_node* pb_gnode = iter->second;
+        VTR_ASSERT(pb_gnode != nullptr);
+        return pb_gnode;
+    }
+
+    /**
+     * @brief Returns the total number of molecules in the prepacker.
+     */
+    inline size_t get_num_molecules() const {
+        size_t num_molecules = 0;
+        t_pack_molecule* molecule_head = list_of_pack_molecules;
+        for (auto cur_molecule = molecule_head; cur_molecule != nullptr; cur_molecule = cur_molecule->next) {
+            ++num_molecules;
+        }
+        return num_molecules;
+    }
+
+    /**
+     * @brief Returns all of the molecules as a vector.
+     */
+    inline std::vector<t_pack_molecule*> get_molecules_vector() const {
+        std::vector<t_pack_molecule*> molecules;
+        t_pack_molecule* molecule_head = list_of_pack_molecules;
+        for (auto cur_molecule = molecule_head; cur_molecule != nullptr; cur_molecule = cur_molecule->next) {
+            molecules.push_back(cur_molecule);
+        }
+        return molecules;
+    }
+
+    /**
+     * @brief Marks all of the molecules as valid.
```

Comment:
If you know what valid means in this context you could define it here.

Reply:
I'll give it a look; honestly I am not sure, and this is the only method that mutates the prepacker (all others are read only, other than init obviously). I am also curious.